File: main/video.js
// ============================================================
// 主进程视频服务：WebM/WAV 离线合成 MP4
// ============================================================
'use strict';

function registerVideoIpc({ ipcMain, dialog, BrowserWindow, app, path, fs, runEngineInline, parsePyJson }) {
  // 视频导出：接收渲染器录制的 WebM（可视化 + 可选离线渲染的 WAV 音频）→ 内置 ffmpeg 合成 MP4
  ipcMain.handle('video:transcode', async (evt, opts) => {
    try {
      if (!opts || !opts.data) return { ok: false, error: 'empty' };
      const win = BrowserWindow.fromWebContents(evt.sender);
      const save = await dialog.showSaveDialog({
        title: '导出视频 MP4',
        defaultPath: path.join(app.getPath('downloads'), 'FuFumidi-video.mp4'),
        filters: [{ name: 'MP4 视频', extensions: ['mp4'] }],
      });
      if (save.canceled || !save.filePath) return { ok: false, canceled: true };
      const tmpDir = path.join(app.getPath('temp'), 'fufumidi-video');
      fs.mkdirSync(tmpDir, { recursive: true });
      const webm = path.join(tmpDir, Date.now() + '.webm');
      fs.writeFileSync(webm, Buffer.from(opts.data));
      let wav = null;
      if (opts.audio && opts.audio.byteLength) {
        wav = path.join(tmpDir, Date.now() + '.wav');
        fs.writeFileSync(wav, Buffer.from(opts.audio));
      }
      const out = save.filePath;
      const args = ['-y', '-hide_banner', '-loglevel', 'error', '-i', webm];
      if (wav) args.push('-i', wav, '-map', '0:v:0', '-map', '1:a:0');
      args.push('-c:v', 'libx264', '-pix_fmt', 'yuv420p', '-c:a', 'aac', '-shortest', '-movflags', '+faststart', out);
      const code =
        'import json, subprocess\n' +
        'import imageio_ffmpeg\n' +
        'ff = imageio_ffmpeg.get_ffmpeg_exe()\n' +
        'r = subprocess.run([ff] + ' + JSON.stringify(args) +
        ', capture_output=True)\n' +
        "print(json.dumps({'ok': r.returncode == 0, 'err': (r.stderr or b'').decode('utf-8', 'replace')[-300:]}))";
      const rr = await runEngineInline(code);
      try { fs.unlinkSync(webm); if (wav) fs.unlinkSync(wav); } catch (e) {}
      const d = parsePyJson(rr.out);
      if (d && d.ok) return { ok: true, path: out };
      return { ok: false, error: (d && d.err) || (rr.out || 'ffmpeg failed').slice(-300) };
    } catch (e) { return { ok: false, error: String((e && e.message) || e) }; }
  });
}

module.exports = { registerVideoIpc };

File: src-tauri/src/video.rs
// 视频导出：WebM（可视化）+ 可选 WAV → ffmpeg 合成 MP4 —— 移植自 main/video.js
use serde::Deserialize;
use serde_json::{json, Value};
use std::fs;
use tauri::{AppHandle, Manager};
use tauri_plugin_dialog::DialogExt;

use crate::engine::{parse_py_json, run_inline_python};

#[derive(Deserialize)]
pub struct TranscodeOpts {
    data: Option<Vec<u8>>,
    audio: Option<Vec<u8>>,
}

// 合成 MP4：保存对话框 → 写临时 WebM/WAV → Python imageio_ffmpeg 编码
#[tauri::command]
pub async fn transcode_video(
    app: AppHandle,
    window: tauri::WebviewWindow,
    opts: Option<TranscodeOpts>,
) -> Value {
    let opts = match opts {
        Some(o) => o,
        None => return json!({ "ok": false, "error": "empty" }),
    };
    let webm_data = match opts.data {
        Some(d) if !d.is_empty() => d,
        _ => return json!({ "ok": false, "error": "empty" }),
    };
    let saved = window
        .dialog()
        .file()
        .add_filter("MP4 视频", &["mp4"])
        .set_file_name("FuFumidi-video.mp4")
        .blocking_save_file();
    let out_path = match saved.and_then(|f| f.into_path().ok()) {
        Some(p) => p,
        None => return json!({ "ok": false, "canceled": true }),
    };

    let tmp_dir = app
        .path()
        .app_cache_dir()
        .unwrap_or_else(|_| std::path::PathBuf::from("."))
        .join("fufumidi-video");
    let _ = fs::create_dir_all(&tmp_dir);
    let stamp = std::time::SystemTime::now()
        .duration_since(std::time::UNIX_EPOCH)
        .map(|d| d.as_millis())
        .unwrap_or(0);
    let webm = tmp_dir.join(format!("{}.webm", stamp));
    let _ = fs::write(&webm, &webm_data);
    let mut wav: Option<std::path::PathBuf> = None;
    if let Some(a) = &opts.audio {
        if !a.is_empty() {
            let w = tmp_dir.join(format!("{}.wav", stamp + 1));
            let _ = fs::write(&w, a);
            wav = Some(w);
        }
    }

    let out = out_path.to_string_lossy().to_string();
    let webm_s = webm.to_string_lossy().to_string();
    let wav_s = wav.as_ref().map(|w| w.to_string_lossy().to_string());
    let mut ff_args: Vec<String> = vec![
        "-y".into(),
        "-hide_banner".into(),
        "-loglevel".into(),
        "error".into(),
        "-i".into(),
        webm_s.clone(),
    ];
    if let Some(w) = &wav_s {
        ff_args.push("-i".into());
        ff_args.push(w.clone());
        ff_args.push("-map".into());
        ff_args.push("0:v:0".into());
        ff_args.push("-map".into());
        ff_args.push("1:a:0".into());
    }
    ff_args.extend([
        "-c:v".into(),
        "libx264".into(),
        "-pix_fmt".into(),
        "yuv420p".into(),
        "-c:a".into(),
        "aac".into(),
        "-shortest".into(),
        "-movflags".into(),
        "+faststart".into(),
        out.clone(),
    ]);
    let args_json = serde_json::to_string(&ff_args).unwrap_or_else(|_| "[]".into());
    let code = format!(
        "import json, subprocess\nimport imageio_ffmpeg\nff = imageio_ffmpeg.get_ffmpeg_exe()\nr = subprocess.run([ff] + json.loads(r'''{}'''), capture_output=True)\nprint(json.dumps({{'ok': r.returncode == 0, 'err': (r.stderr or b'').decode('utf-8', 'replace')[-300:]}}))",
        args_json
    );
    let r = run_inline_python(&app, &code);
    let _ = fs::remove_file(&webm);
    if let Some(w) = wav {
        let _ = fs::remove_file(w);
    }
    match parse_py_json(&r.out) {
        Some(d) if d.get("ok").and_then(|x| x.as_bool()).unwrap_or(false) => {
            json!({ "ok": true, "path": out })
        }
        Some(d) => json!({ "ok": false, "error": d.get("err").and_then(|x| x.as_str()).unwrap_or("ffmpeg failed") }),
        None => json!({ "ok": false, "error": r.out.chars().take(300).collect::<String>() }),
    }
}
